javascript/documentation/library-customization.rst

```diff
@@ -0,0 +1,249 @@
+
+Type tracking
+~~~~~~~~~~~~~
+
+It is not normally necessary to customize this layer.
+
+Framework models
+~~~~~~~~~~~~~~~~
+
+The ``semmle.javascript.frameworks.HTTP`` module defines many abstract classes that can be extended
+to implement support for new web frameworks. These classes, in turn, are used by some of the
+security queries (such as the cross-site scripting queries) to define sources and sinks, so these
+queries will automatically benefit from the additional modeling.
+
+Similarly, the ``semmle.javascript.frameworks.SQL`` module defines abstract classes for modeling SQL
+connector libraries, and the ``semmle.javascript.JsonParsers`` and
+``semmle.javascript.frameworks.XML`` modules for modeling JSON and XML parsers, respectively.
+
+The ``semmle.javascript.Concepts`` modules defines a few very broad concepts such as system-command
+executions or file-system accesses, which are concretely instantiated in some of the existing
+framework libraries, but can of course be further extended to model additional frameworks.
+
+Global data flow and taint tracking
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Most security queries consist of one QL file defining the query, one configuration module defining
+the taint-tracking configuration, and one customization module defining sources, sinks and
+sanitizers. For example, ``Security/CWE-078/CommandInjection.ql`` defines the command-injection
+query. It imports module ``semmle.javascript.security.dataflow.CommandInjection``, which defines the
+configuration class ``CommandInjection::Configuration``, and itself imports module
+``semmle.javascript.security.dataflow.CommandInjectionCustomizations``, which defines sources, sinks
+and sanitizers by means of three abstract classes ``CommandInjection::Source``,
+``CommandInjetion::Sink`` and ``CommandInjection::Sanitizer``, respectively.
+
+To define additional sources, sinks or sanitizers for this or any other security query, import the
+customization module and extend these abstract classes with additional subclasses.
+
+Note that you should normally only import the configuration module from a QL file. Importing it into
+the standard library (for example by importing it in ``Customizations.qll``) will slow down all the
+other security queries, since the configuration class will now be always in scope and flow from its
+sources to sinks will be tracked in addition to all the other configuration classes.
+
+Another useful extension point is the class ``RemoteFlowSource``, which is used as a source by most
+queries looking for injection vulnerabilities (such as SQL injection or cross-site scripting). By
+extending it with new subclasses modelling other sources of user-controlled input you can
+simultaneously improve all of these queries.
+
+Finally, you can extend the classes ``Dataflow::AdditionalSource``, ``DataFlow::AdditionalSink``,
```

Comment:
This requires a repeat of the warning about importing multiple configurations by default.

I actually do not remember seeing those `Additional*` classes before. Do we still want to advertise them now that we have the `*Customizations.qll` files?

Reply:
No, probably fine not to mention them.